src/rewrite_reduce.cpp

```diff
@@ -0,0 +1,109 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#include <migraphx/rewrite_reduce.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/matcher.hpp>
+#include <migraphx/make_op.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+namespace {
+struct find_softmax
+{
+    auto matcher() const { return match::name("softmax"); }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        auto ins  = r.result;
+        auto op   = ins->get_operator().to_value();
+        auto axis = op["axis"].to<std::int64_t>();
+
+        auto input = ins->inputs().front();
+        auto max   = m.insert_instruction(ins, make_op("reduce_max", {{"axes", {axis}}}), input);
+        auto maxb  = m.insert_instruction(
+            ins, make_op("multibroadcast", {{"out_lens", input->get_shape().lens()}}), max);
+        auto sub  = m.insert_instruction(ins, make_op("sub"), input, maxb);
+        auto exp  = m.insert_instruction(ins, make_op("exp"), sub);
+        auto sum  = m.insert_instruction(ins, make_op("reduce_sum", {{"axes", {axis}}}), exp);
+        auto sumb = m.insert_instruction(
+            ins, make_op("multibroadcast", {{"out_lens", input->get_shape().lens()}}), sum);
+        m.replace_instruction(ins, make_op("div"), exp, sumb);
+    }
+};
+
+struct find_reduce_mean
+{
+    auto matcher() const { return match::name("reduce_mean"); }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        auto reduce_mean = r.result;
+        auto op          = reduce_mean->get_operator().to_value();
+        auto axes        = op["axes"].to_vector<std::int64_t>();
+        auto input       = reduce_mean->inputs().front();
+
+        bool is_integral = false;
+        double max_n     = 0;
+        input->get_shape().visit_type([&](auto t) {
+            is_integral = t.is_integral();
+            max_n       = t.max();
+        });
+
+        // avoid integer division
+        if(is_integral)
+            return;
+
+        auto dims = input->get_shape().lens();
+        double n  = std::accumulate(
+            axes.begin(), axes.end(), 1, [&](const auto& prod_value, const auto& axis) {
+                auto norm_axis = axis < 0 ? dims.size() + axis : axis;
+                return prod_value * dims.at(norm_axis);
+            });
+
+        // avoid overflow
+        if(n >= max_n)
```

Comment:
Ideally, we should use a higher precision for these cases. For now you could just change it to `n >= max_n / 4`(so the logic [here](https://github.com/ROCm/AMDMIGraphX/blob/develop/src/targets/gpu/compile_gen.cpp#L227) will still be applied).